### ISSUE TITLE
Stuck while building

### ISSUE THREAD

**Issue by InternetFreedom** [cisco/libacvp#592]
Steps executed:
1. ./configure LIBS=-ldl -with-ssl-dir=/usr/local/ssl --disable-kdf --enable-static --enable-offline --with-fom-dir=/home/security/canisters-fips/Debian/openssl/fips
2. make
I get stuck while building app_main.c:
make[1]: Entering directory '/home/security/libacvp/app'
/home/security/canisters-fips/Debian/openssl/fips/bin/fipsld -DPACKAGE_NAME=\"libacvp\" -DPACKAGE_TARNAME=\"libacvp\" -DPACKAGE_VERSION=\"1.1.3\" -DPACKAGE_STRING=\"libacvp\ 1.1.3\" -DPACKAGE_BUGREPORT=\"https://github.com/cisco/libacvp/issues\" -DPACKAGE_URL=\"\" -DPACKAGE=\"libacvp\" -DVERSION=\"1.1.3\" -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -I.    -g -fPIE -I../include -I/usr/local/ssl/include -DACVP_NO_RUNTIME -DOPENSSL_FIPS -I/home/security/canisters-fips/Debian/openssl/fips/include -I/home/security/libacvp/safe_c_stub/include -g -O2 -Wall -fstack-protector-all -D_FORTIFY_SOURCE=2 -Wl,-z,noexecstack -DACVP_OFFLINE -MT acvp_app-app_main.o -MD -MP -MF .deps/acvp_app-app_main.Tpo -c -o acvp_app-app_main.o `test -f 'app_main.c' || echo './'`app_main.c

Thanks in advance for your help

**Comment by abkarcher** [comment]
Hello,

I find i run into this issue with static builds unless this is set in my environment - 
`export FIPSLD_CC=gcc`
Then re-configure + make clean and try again.

Hope this helps,
Andrew

**Comment by InternetFreedom** [comment]
Hi Andrew,
Thanks for the quick response. That fixed the issue but my make failed at 
libtool: link: /home/security/canisters-fips/Debian/openssl/fips/bin/fipsld -g -fPIE -I../include -I/usr/local/ssl/include -DACVP_NO_RUNTIME -DOPENSSL_FIPS -I/home/security/canisters-fips/Debian/openssl/fips/include -I/home/security/libacvp/safe_c_stub/include -g -O2 -Wall -fstack-protector-all -D_FORTIFY_SOURCE=2 -Wl,-z -Wl,noexecstack -DACVP_OFFLINE -o acvp_app acvp_app-app_main.o acvp_app-app_aes.o acvp_app-app_cli.o acvp_app-app_cmac.o acvp_app-app_des.o acvp_app-app_drbg.o acvp_app-app_dsa.o acvp_app-app_ecdsa.o acvp_app-app_hmac.o acvp_app-app_kas.o acvp_app-app_kdf.o acvp_app-app_rsa.o acvp_app-app_sha.o acvp_app-app_utils.o /home/security/canisters-fips/Debian/openssl/fips/lib/fipscanister.o  -L/home/security/libacvp/src/.libs /home/security/libacvp/src/.libs/libacvp.a -L/usr/local/ssl/lib -lcrypto -L/home/security/canisters-fips/Debian/openssl/fips/lib -ldl
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/9/../../../x86_64-linux-gnu/Scrt1.o: in function `_start':
(.text+0x24): undefined reference to `main'
collect2: error: ld returned 1 exit status

Do you have any idea why I am getting a link error?

Thank you

**Comment by InternetFreedom** [comment]
HI @abkarcher,
I do not know what is causing the 'undefined reference to main error'. The steps I tooks are:
1. export FIPSLD_CC=gcc
2. ./configure --with-ssl-dir=/usr/local/ssl  --enable-offline --with-fom-dir=/home/security/canisters-fips/Debian/openssl/fips  --disable-kdf
2. make clean
3. make

Thanks for your help!

**Comment by abkarcher** [comment]
Hello,

Apologies for the delay, I was out of the office. have you had any further luck with this? I am unable to replicate your issue using any combination of arguments currently. A few questions to maybe help me replicate,

What version of libacvp are you trying to build?
Have you made any modifications to app_main.c?
What version of OpenSSL/FOM are you using?
I do see some oddities with ./configure, openSSL fom, and --enable-offline. Are you manually setting anything to LDFLAGS, LIBS, or linker-related arguments in CFLAGS? Are you still using LIBS=-ldl for configure the second time as well?

(If you are getting code from master, I recommend using the releases or the highest-number throttle branch instead).

Thanks,
Andrew

**Comment by InternetFreedom** [comment]
Hi @abkarcher,

I am still having the same issue. The steps I took are:
1. export FIPSLD_CC=gcc
2. ./configure LIBS=-ldl --with-ssl-dir=/usr/local/ssl --enable-static --enable-offline --with-fom-dir=/home/security/canisters-fips/Debian/openssl/fips
3. make clean
4. make

I am currently running libacvp 1_4_0 throttle. Openssl version is 1.0.2l and FOM version is 2.0.6.

Error message:
libtool: link: /home/security/canisters-fips/Debian/openssl/fips/bin/fipsld -g -fPIE -I../include -I/usr/local/ssl/include -DACVP_NO_RUNTIME -DOPENSSL_FIPS -DOPENSSL_KDF_SUPPORT -I/home/security/canisters-fips/Debian/openssl/fips/include -I/home/security/libacvp/safe_c_stub/include -g -O2 -Wall -fstack-protector-all -D_FORTIFY_SOURCE=2 -Wl,-z -Wl,noexecstack -DACVP_OFFLINE -o acvp_app acvp_app-app_main.o acvp_app-app_aes.o acvp_app-app_cli.o acvp_app-app_cmac.o acvp_app-app_des.o acvp_app-app_drbg.o acvp_app-app_dsa.o acvp_app-app_ecdsa.o acvp_app-app_hmac.o acvp_app-app_kas.o acvp_app-app_kdf.o acvp_app-app_rsa.o acvp_app-app_sha.o acvp_app-app_utils.o /home/security/canisters-fips/Debian/openssl/fips/lib/fipscanister.o  -L/home/security/libacvp/src/.libs /home/security/libacvp/src/.libs/libacvp.a -L/usr/local/ssl/lib -lcrypto -L/home/security/canisters-fips/Debian/openssl/fips/lib -ldl
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/9/../../../x86_64-linux-gnu/Scrt1.o: in function `_start':
(.text+0x24): undefined reference to `main'
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:437: acvp_app] Error 1
make[1]: Leaving directory '/home/security/libacvp/app'
make: *** [Makefile:454: all-recursive] Error 1

**Comment by InternetFreedom** [comment]
Hi @abkarcher,
I have been able to fix the issue. 
I tried testing out acvp_app with this command: 
./app/acvp_app --all_algs --vector_req ../ACVP-AES-CBC-request.json --vector_rsp ../ACVP-AES-CBC-response.json 
 and this is the output I got:

Using the following parameters:

    ACV_SERVER:     127.0.0.1
    ACV_PORT:       443
    ACV_URI_PREFIX: /acvp/v1/

***ACVP [STATUS][acvp_run_vectors_from_file:880]--> Beginning offline processing of vector sets...
***ACVP [WARN][acvp_run_vectors_from_file:914]--> Missing session URL, results will not be POSTed to server

since I am testing offline, why is it trying to post my response to the server? I will like the response to be saved on my local system. How can I get the response locally instead of posting it to the server?
Thank you

**Comment by abkarcher** [comment]
Hello,

The session URL is saved in the request file, and then copied from the request file to the response file, so when --vector_upload is run, it uploads the response file to the URL in the file.

If you are running offline, it still should be able to process the test results and place them in the response file, but running acvp_app with --vector_upload will not know what URL to use. 

Was the session URL manually removed from the JSON file, or was the JSON file acquired elsewhere? this way of saving the URLs is specific to libacvp.

Thanks,
Andrew

**Comment by InternetFreedom** [comment]
Hi @abkarcher ,

I do not see the response file anywhere. It is not in the path I specified, will it be saved somewhere? I got the request file from our testing lab and not directly from the demo server. My company is using a testing lab for our certification and they provided the JSON request files.

 I installed the FOM on my device and used this command to configure:
./configure --with-ssl-dir=/usr/local/ssl --with-fom-dir=/usr/local/ssl/fips-2.0 --disable-kdf --enable-static --enable-offline LIBS=-ldl
This seem to have fixed it. 

Thank you
